go/libraries/doltcore/remotestorage/internal/ranges/ranges.go

```diff
@@ -0,0 +1,252 @@
+// bytes.
+//
+// A |GetRange| struct is a member of a |Region| in the |RegionHeap|.
+type GetRange struct {
+	Url    string
+	Hash   []byte
+	Offset uint64
+	Length uint32
+	Region *Region
+}
+
+// A |Region| represents a continuous range of bytes within in a Url.
+// |ranges.Tree| maintains |Region| instances that cover every |GetRange|
+// within the tree. As entries are inserted into the Tree, their Regions can
+// coallesce with Regions which come before or after them in the same Url,
+// based on the |coallesceLimit|.
+//
+// |Region|s are maintained in a |RegionHeap| so that the |Tree| can quickly
+// return a large download to get started on when a download worker is
+// available.
+type Region struct {
+	Url          string
+	StartOffset  uint64
+	EndOffset    uint64
+	MatchedBytes uint64
+	HeapIndex    int
+	Score        int
+}
+
+type RegionHeap []*Region
+
+func (rh RegionHeap) Len() int {
+	return len(rh)
+}
+
+const (
+	HeapStrategy_smallest = iota
+	HeapStrategy_largest
+	HeapStrategy_random
+)
+
+var strategy = HeapStrategy_largest
+
+func (rh RegionHeap) Less(i, j int) bool {
+	leni := rh[i].EndOffset - rh[i].StartOffset
+	lenj := rh[j].EndOffset - rh[j].StartOffset
+	if strategy == HeapStrategy_largest {
+		// This makes us track the largest region...
+		return leni > lenj
+	} else if strategy == HeapStrategy_smallest {
+		// This makes us track the smallest...
+		return leni < lenj
+	} else {
+		// This makes us track a random order...
+		return rh[i].Score < rh[j].Score
+	}
+}
+
+func (rh RegionHeap) Swap(i, j int) {
+	rh[i], rh[j] = rh[j], rh[i]
+	rh[i].HeapIndex = i
+	rh[j].HeapIndex = j
+}
+
+func (rh *RegionHeap) Push(x any) {
+	r := x.(*Region)
+	*rh = append(*rh, r)
+	r.HeapIndex = len(*rh) - 1
+}
+
+func (rh *RegionHeap) Pop() any {
+	old := *rh
+	n := len(old)
+	r := old[n-1]
+	*rh = old[0 : n-1]
+	return r
+}
+
+// A ranges.Tree is a tree data structure designed to support efficient
+// coallescing of non-overlapping ranges inserted into it.
+type Tree struct {
+	t              *btree.BTreeG[*GetRange]
+	regions        *RegionHeap
+	coallesceLimit int
+}
+
+func GetRangeLess(a, b *GetRange) bool {
+	if a.Url == b.Url {
+		return a.Offset < b.Offset
+	} else {
+		return a.Url < b.Url
+	}
+}
+
+func NewTree(coallesceLimit int) *Tree {
+	return &Tree{
+		t:              btree.NewG[*GetRange](64, GetRangeLess),
+		regions:        &RegionHeap{},
+		coallesceLimit: coallesceLimit,
+	}
+}
+
+// |intern| will deduplicate strings that are stored in the |ranges.Tree|, so
+// that all equal values share the same heap memory. The context is that URLs
+// stored in the |Tree| can be very long, since they can be pre-signed S3 URLs,
+// for example. And in general a Tree will have a large number of |GetRange|
+// entries, that contain the same |Url|.
+func (t *Tree) intern(s string) string {
+	t.t.AscendGreaterOrEqual(&GetRange{Url: s}, func(gr *GetRange) bool {
+		if gr.Url == s {
+			s = gr.Url
```

Comment:
Hmm...how can I make it more clear? The intent here is that there are `*GetRange` instances stored in the tree, and those instances might have a `Url` matching the incoming `Url`. Essentially, we want to take the existing `StringHeader` of an existing entry if it exists, because that will share backing storage with something that's already in the tree...so, we're assigning `s` here, since it's the thing we're going to return down below.

Reply:
yeah ig that makes sense, forward-looking sharing vs backward

Reply:
Yeah, the thing about forward sharing is that there will be an arbitrary, possibly very large, number of entries already in the tree with `Url == s`